push_notifications/apns.py

```diff
@@ -4,100 +4,165 @@
 https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html
 """
 
+import asyncio
+import contextlib
+import tempfile
 import time
 
-from apns2 import client as apns2_client
-from apns2 import credentials as apns2_credentials
-from apns2 import errors as apns2_errors
-from apns2 import payload as apns2_payload
+import aioapns
+from aioapns.common import APNS_RESPONSE_CODE, PRIORITY_HIGH, PRIORITY_NORMAL
+from asgiref.sync import async_to_sync
 
 from . import models
 from .conf import get_manager
-from .exceptions import APNSError, APNSUnsupportedPriority, APNSServerError
-
+from .exceptions import APNSError, APNSServerError, APNSUnsupportedPriority
+
+
+SUCCESS_RESULT = "Success"
+UNREGISTERED_RESULT = "Unregistered"
+
+
+@contextlib.contextmanager
+def _apns_path_for_cert(cert):
+	if cert is None:
+		yield None
+	with tempfile.NamedTemporaryFile("w") as cert_file:
+		cert_file.write(cert)
+		cert_file.flush()
+		yield cert_file.name
+
+
+def _apns_create_client(application_id=None):
+	cert = None
+	key_path = None
+	key_id = None
+	team_id = None
+
+	if not get_manager().has_auth_token_creds(application_id):
+		cert = get_manager().get_apns_certificate(application_id)
+	else:
+		key_path, key_id, team_id = get_manager().get_apns_auth_creds(application_id)
+		# No use getting a lifetime because this credential is
+		# ephemeral, but if you're looking at this to see how to
+		# create a credential, you could also pass the lifetime and
+		# algorithm. Neither of those settings are exposed in the
+		# settings API at the moment.
+
+	with _apns_path_for_cert(cert) as cert_path:
+		client = aioapns.APNs(
+			client_cert=cert_path,
+			key=key_path,
+			key_id=key_id,
+			team_id=team_id,
+			use_sandbox=get_manager().get_apns_use_sandbox(application_id),
+		)
```

Comment:
I found two errors. The topic isn't being sent when creating the client and the context processor raises an exception if there's no credential path. I tested it afterwards and works great with Django 4.2 and Python 3.11

`	if not get_manager().has_auth_token_creds(application_id):
		cert = get_manager().get_apns_certificate(application_id)
		with _apns_path_for_cert(cert) as cert_path:
			client = aioapns.APNs(
				client_cert=cert_path,
				team_id=team_id,
				topic=get_manager().get_apns_topic(application_id),
				use_sandbox=get_manager().get_apns_use_sandbox(application_id),
			)
	else:
		key_path, key_id, team_id = get_manager().get_apns_auth_creds(application_id)
		client = aioapns.APNs(
			key=key_path,
			key_id=key_id,
			team_id=team_id,
			topic=get_manager().get_apns_topic(application_id),
			use_sandbox=get_manager().get_apns_use_sandbox(application_id),
		)`

Reply:
Here's a fork with the fix
[[https://github.com/aalbinati/django-push-notifications/tree/remove-apns2](url)](https://github.com/aalbinati/django-push-notifications/tree/remove-apns2)

Reply:
Awesome. Thank you @aalbinati. Doing some testing with the service was on my to-do, I'll pull in the changes you're suggesting soon.